dataland-frontend/src/components/general/apiKey/CreateApiKeyCard.vue

```diff
@@ -52,7 +52,7 @@
           {{
```

Comment:
but what if (expireTimeDropdown === "custom") // true , but (this.expireTimeDays > 0) // false ?

Reply:
Then it will enter the last "else" block.

I think you could simplify the whole function to:

```TypeScript
checkDateAndEmitGenerateApiKey() {
      if (
        (this.expireTimeDropdown === "custom" && this.expireTimeDays > 0)
      ) {
        this.$emit("generateApiKey", this.expireTimeDays);
      } else if (this.expireTimeDropdown === "noExpiry") {
        this.$emit("generateApiKey");
      } else {
        this.isExpireTimeCorrect = false;
      }
    },
```

If you adjust the ```setExpireTimeDays()``` to do:
```
if (event.value === "noExpiry") {
        this.expireTimeDays = null
```

you could even write:

```TypeScript
checkDateAndEmitGenerateApiKey() {
      if (
        (this.expireTimeDropdown === "custom" && this.expireTimeDays > 0) 
        | |
        (this.expireTimeDropdown === "noExpiry")
      ) {
        this.$emit("generateApiKey", this.expireTimeDays);
      } else {
        this.isExpireTimeCorrect = false;
      }
    },
```

Reply:
Just some ideas from reading the code.
Haven't checked if those codes actually work.

Reply:
refactored

Reply:
If we enforce the Calendar to only allow dates in the future (see my other comment), we could even further simplify this method to:

```typescript
checkDateAndEmitGenerateApiKey() {
      if (this.expireTimeDays) {
        this.$emit("generateApiKey", this.expireTimeDays);
      } else if (this.expireTimeDropdown === "noExpiry") {
        this.$emit("generateApiKey");
      } else {
        this.isExpiryDateValid = false;
      }
    },
```

Reply:
I think you can even write:

```typescript
checkDateAndEmitGenerateApiKey() {
      if (this.expireTimeDays || (this.expireTimeDropdown === "noExpiry") {
        this.$emit("generateApiKey", this.expireTimeDays);
     else {
        this.isExpiryDateValid = false;
      }
    },
```

Reason:   If a null value is passed as ```daysValid``` to the ```generateApiKey()``` method, it automatically recognizes that and sets the expiryDate of the api key to null.

Reply:
but doesn't the function expect a number or undefined ?  " generateApiKey(daysValid?: number, ... " 
" Argument of type 'null' is not assignable to parameter of type 'number | undefined'. "

Reply:
Yes but if we write it as:

```typescript
async generateApiKey(expirationTime?: number | null) {
      try {
        this.waitingForData = true;
        const keycloakPromiseGetter = assertDefined(this.getKeycloakPromise);

        const apiKeyManagerController = await new ApiClientProvider(
          keycloakPromiseGetter()
        ).getApiKeyManagerController();
        let response
        if(expirationTime) {
          response = await apiKeyManagerController.generateApiKey(expirationTime);
        } else {
          response = await apiKeyManagerController.generateApiKey();
        }
        this.waitingForData = false;
        this.existsApiKey = true;
        this.expiryDate = response.data.apiKeyMetaInfo.expiryDate ? response.data.apiKeyMetaInfo.expiryDate : 0;
        this.newKey = response.data.apiKey;
        this.userRolesAccordingToApiKey = response.data.apiKeyMetaInfo.keycloakRoles
          ? response.data.apiKeyMetaInfo.keycloakRoles
          : [];
        this.setActivePageState("view");
      } catch (error) {
        console.error(error);
        this.existsApiKey = false;
      }
    },
```

Then it can also handle null values.

Reply:
Yes I can call function generateApiKey with null param but in " ApiKeyControllerApiInterface " function:
generateApiKey(daysValid?: number, options?: AxiosRequestConfig): AxiosPromise<ApiKeyAndMetaInfo>;

daysValid is type number | undefined right ?

Reply:
Yes it is. But ```daysValid``` is also optional in the ApiKeyControllerApiInterface function generateApiKey.
That means you can just check if expiration type is null, and in that case not pass a daysValid value to the function:

```typescript
if(expirationTime) {
          response = await apiKeyManagerController.generateApiKey(expirationTime);
        } else {
          response = await apiKeyManagerController.generateApiKey();
        }
        ```

Reply:
But...aren't we doing just that in this condition:
} else if (this.expiryTimeDropdown === "noExpiry") {
         this.$emit("generateApiKey");
         ??

Reply:
Well at the end it's a matter of taste/coding philosophy. 

1. Option (which is currently implemented) is using 
} else if (this.expiryTimeDropdown === "noExpiry") {
this.$emit("generateApiKey");
and therefore sending an undefined value for "daysValid" in case the user picks "noExpiry".
So it assures that only values are emitted, that can be handled by the ```generateApiKey()``` function in the ApiKeysPage.

2. Option is that in case of "noExpiry", the emit is ```this.$emit("generateApiKey", this.expireTimeDays)``` with a ```null``` inside of the emitted value.
In this case the ```generateApiKey()``` function in the ApiKeysPage.vue needs to check for null value like in my last comment and handle it.

Bot are fully working and ok from a functionality point of view.
The motivation behind my suggestion was just that the "CreateApiKeyCard.vue" component should not care about which values are allowed to be emitted and which aren't. If the user picks "noExpiry", then it should be ok for it to emit the null value.
It is the ```generateApiKey()``` functions responsibility to live with that emit.
So it was more out of the idea of a "seperation of responsibilities".

The more importnat thing I wanted to discuss with you in this conversation anyways is that the validityCheck can be omitted if we enforce the date picker to only allow future dates 😃

Reply:
It is ok for you if we leave condition like it is and check attribute before we call function generate Api Key rather than check it insite function itself ?

Reply:
Yes it's ok.